Fix "Not Participating" filter checkbox on Discover Challenges screen, Fixes #13025

lizapritchard opened the pull request:
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #13025

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

- Filtering added on the front-end of the "discover challenges" view,  to allow the user to view challenges they _are_ and _are not_ currently participating in. Mimics function in "my challenges" view. 
- Removed "role" filter from "my challenges" and "discover challenges" views (deemed extraneous by @saraolson in #13025 followups)
<img width="1680" alt="Screen Shot 2021-04-21 at 10 26 30 AM" src="https://user-images.githubusercontent.com/49509154/115570465-241e0380-a28c-11eb-995c-1103fbfa2bdd.png">
<img width="1680" alt="Screen Shot 2021-04-21 at 10 26 41 AM" src="https://user-images.githubusercontent.com/49509154/115570475-26805d80-a28c-11eb-9d0b-4e871944fcd4.png">





[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID:  6179e6e4-1070-43a1-b1df-c1aad12a3ef7

lizapritchard committed:
change 'role' options and filters to 'membership'

lizapritchard committed:
added membership filtering function to discover challenges

lizapritchard committed:
fix accidental deletion

SabreCat commented on website/client/src/components/challenges/findChallenges.vue:
```diff
@@ -171,7 +171,23 @@ export default {
     ...mapState({ user: 'user.data' }),
     filteredChallenges () {
-      return this.challenges;
+      const { filters } = this;
```
The code is identical between the two Challenge Vue files, now--could this be abstracted as a mixin?

SabreCat replied:
We already have the `challengeUtilities` mixin, so it could live there

SabreCat commented on website/client/src/components/challenges/sidebar.vue:
```diff
@@ -34,7 +34,7 @@
         </div>
       </filter-group>
-      <filter-group :title="$t('role')">
+      <!-- <filter-group :title="$t('role')">
```
OK to delete these rather than commenting out. Git history will preserve it if we need to restore it!

SabreCat commented on website/client/src/components/challenges/sidebar.vue:
```diff
@@ -231,6 +244,7 @@ export default {
         categories: this.categoryFilters,
         roles: this.roleFilters,
```
We can remove role filtering here too

SabreCat reviewed (changes requested):
Thanks for your help here! See minor comments in review, but this shouldn't be difficult to move across the finish line!

SabreCat commented:
One functional thing I'm seeing: selecting _both_ Participating and Not Participating causes the filter to show only the Not Participating challenges. Compare the behavior of the Owned and Not Owned, which if you tick both acts the same as if neither is checked, i.e., not excluding any Challenges from the current list.

lizapritchard committed:
complete removal of roles from challenge filters

lizapritchard committed:
abstract challenge "participating" filtering into mixin

lizapritchard committed:
selecting "participating" and "not participating" shows all challenges

lizapritchard commented:
@SabreCat thanks for the feedback! I think I addressed all of these issues in the new commits, but I was a little unsure about the mixin stuff so there might be some style conventions here that I'm missing

SabreCat commented:
Looks great! Thanks for the updates! We'll get this into the staging environment and from there to production. Credited toward your second contributor tier!
